pylint/checkers/exceptions.py

```diff
@@ -254,6 +255,15 @@ class ExceptionsChecker(checkers.BaseChecker):
 
     def open(self) -> None:
         self._builtin_exceptions = _builtin_exceptions()
+        if any(
+            "." in exc_name for exc_name in self.linter.config.overgeneral_exceptions
+        ):
+            warnings.warn(
```

Comment:
I think the issue you refer to is that `DeprecationWarnings` are not normally shown. You could check with `pytest` if they do show. A test for this would be good.

Reply:
Using `python -Wdefault -m pylint ...` will make them show up but the problem is that absolutely nobody runs pylint like that (and they probably shouldn't, which I'll mention below).
People can see deprecation warnings when something they use in their own code is deprecated because their testing framework will enable deprecation warnings for them and so when the code is tested, they should see it. This is, however, not the case for running pylint because you don't run pylint under pytest but rather run it as its own program.

So IMO a proper solution here would be to either enable deprecation warnings for whole pylint (which has a very big problem of users seeing pylint's own deprecation warnings so I don't think it's actually a good solution) or have pylint show this warning by printing it or reporting it as a message, e.g. how something like `unrecognized-option` or `deprecated-pragma` works:
```
************* Module /home/ubuntu/work/pylint/pylintrc
pylintrc:1:0: E0015: Unrecognized option found: overgeneral-exceptions-typoed (unrecognized-option)
```
I think we would probably want it to show under `pylintrc` as `unrecognized-option` does *but* have it be under the `I` category as `deprecated-pragma`?

Reply:
Shall we use `UserWarning` instead? Or just print something to `stdout`? Using the `add_message` mechanism is a bit of a chore as we would to either 1) inject code specific to this option into the configuration parsing or 2) retrieve the place where the option was specified during a run (configuration file, and its name, vs. the CLI). We don't have a mechanism to do this currently.

Reply:
Maybe just print something to stdout? `UserWarning` can at best look like this:
```
Command line or configuration file:1: UserWarning: Specifying exception names in overgeneral-exceptions option without module name has been deprecated. Use fully qualified name (e.g. builtins.Exception) instead.
```
As the format of warnings is pre-defined based on `category`, `filename`, `lineno`, and `message` as passed to `warnings.warn_explicit()` (low-level version of `warnings.warn` which would allow us to at least specify custom `filename` and `lineno` and in the process also prevent code from showing up).

While with stdout we're free to format it however we like (which brings me to the next question, how do we want to format this?) and I'm actually not sure whether using `warnings` module would have any benefits over stdout in this case.

Also as I was typing this, I just thought of one more problem both of these run into... They break `--output-format=json` :grimacing: Maybe using `add_message` is inevitable?

Reply:
Does `UserWarning` also break `json`? That's something to look into for sure then!

But I don't really mind how the `UserWarning` looks. Isn't that what we are going for? Perhaps only add something about it being `pylint` that emits this?

Reply:
> But I don't really mind how the UserWarning looks. Isn't that what we are going for? Perhaps only add something about it being pylint that emits this?

We could have this:
```
pylint: Command line or configuration file:1: UserWarning: Specifying exception names in overgeneral-exceptions option without module name has been deprecated. Use fully qualified name (e.g. builtins.Exception) instead.
```
Or subclass UserWarning:
```
Command line or configuration file:1: PylintUserWarning: Specifying exception names in overgeneral-exceptions option without module name has been deprecated. Use fully qualified name (e.g. builtins.Exception) instead.
```

> Does `UserWarning` also break `json`?

Yep. I feel like neither `UserWarning` nor `print` may end up working here.

Reply:
@Pierre-Sassoulas What do you think about this?

Reply:
Hmm in theory it should be a user warning, but we might need to do it as a breaking change for pylint 3.0 ?

Reply:
What is the normal process? UserWarning after the deprecation or before?

Reply:
I was thinking of two possibilities: 
- UserWarning in 2.X and also 3.X if we modify this in 2.X
- Deprecation warning in 2.X then user warning in 3.X if we do this in 3.X

Reply:
We modify but in a backwards compatible way and we remove that functionality in `3.0`. So I think `UserWarning` makes sense then.

Reply:
What should be done about UserWarning breaking output of `--output-format=json`?
![image](https://user-images.githubusercontent.com/6032823/191978228-d2981542-c8e7-400a-bbfc-7e147b2d71b2.png)

Reply:
Is it really breaking or is it sent to stderr ?

Reply:
It's sent to stderr, I guess it's not breaking then?

Reply:
> It's sent to stderr, I guess it's not breaking then?

That sounds right.

Reply:
Hopefully, this is resolved with the latest 2 commits.